### ISSUE TITLE
createWrapperSync is not a function

### ISSUE THREAD

**Issue by josh-g3** [cbargren/express-brute-mongoose#12]
I'm just trying to do an example app like in the readme. Trace below.

```/Users/joshclark/dev/sandbox/security/rate-limit/express-brute/mongoose/node_modules/mongoose/lib/services/model/applyHooks.js:50
    createWrapperSync('init', objToDecorate.$__init, null, kareemOptions);
    ^

TypeError: schema.s.hooks.createWrapperSync is not a function
    at applyHooks (/Users/joshclark/dev/sandbox/security/rate-limit/express-brute/mongoose/node_modules/mongoose/lib/services/model/applyHooks.js:50:5)
    at Function.compile (/Users/joshclark/dev/sandbox/security/rate-limit/express-brute/mongoose/node_modules/mongoose/lib/model.js:3782:3)
    at Mongoose.model (/Users/joshclark/dev/sandbox/security/rate-limit/express-brute/mongoose/node_modules/mongoose/lib/index.js:381:22)
    at Object.<anonymous> (/Users/joshclark/dev/sandbox/security/rate-limit/express-brute/mongoose/src/server.js:9:20)
    at Module._compile (module.js:660:30)
    at loader (/Users/joshclark/dev/sandbox/security/rate-limit/express-brute/mongoose/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/joshclark/dev/sandbox/security/rate-limit/express-brute/mongoose/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
    at tryModuleLoad (module.js:513:12)
    at Function.Module._load (module.js:505:3)
    at Function.Module.runMain (module.js:701:10)
    at Object.<anonymous> (/Users/joshclark/dev/sandbox/security/rate-limit/express-brute/mongoose/node_modules/babel-cli/lib/_babel-node.js:154:22)
    at Module._compile (module.js:660:30)
    at Object.Module._extensions..js (module.js:671:10)
    at Module.load (module.js:573:32)
    at tryModuleLoad (module.js:513:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! brute-poc@1.0.0 babel-node: `babel-node --presets=env "src/server.js"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the brute-poc@1.0.0 babel-node script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/joshclark/.npm/_logs/2018-01-23T16_01_30_773Z-debug.log
[nodemon] app crashed - waiting for file changes before starting...
/Users/joshclark/dev/sandbox/security/rate-limit/express-brute/mongoose/node_modules/mongoose/lib/services/model/applyHooks.js:50
    createWrapperSync('init', objToDecorate.$__init, null, kareemOptions);
    ^

TypeError: schema.s.hooks.createWrapperSync is not a function
    at applyHooks (/Users/joshclark/dev/sandbox/security/rate-limit/express-brute/mongoose/node_modules/mongoose/lib/services/model/applyHooks.js:50:5)```

**Comment by cbargren** [comment]
Haven't seen that before. What does the code for your schema look like and what version of mongoose are you running in your app?

**Comment by josh-g3** [comment]
Full app code:

```js
import express from 'express';
import ExpressBrute from 'express-brute';
import MongooseStore from 'express-brute-mongoose';
import BruteForceSchema from 'express-brute-mongoose/dist/schema';
import mongoose from 'mongoose';

let
  app = express(),
  model = mongoose.model('bruteforce', BruteForceSchema),
  store = new MongooseStore(model),
  bruteforce = new ExpressBrute(store);

app.post(
  '/auth',
  bruteforce.prevent,
  (req, res, next) => {
    res.send('Success!');
  }
);

app.listen(3000, () => {
  console.info('Express app listening on port 3000...');
});
```

Express v4.16.2
Mongo v3.6.1
Mongoose v5.0.1

e: Updated Mongoose to 5.0.2 and same error. Using nodemon & babel too, if that makes a difference.

**Comment by cbargren** [comment]
Hm... I'm sorry, but I'm not able to repro that locally, even using all those module versions. Could you provide a link to the repository perhaps?

**Comment by josh-g3** [comment]
Just sent you a private repo invite on BB. Thanks for looking into this for me.

**Comment by cbargren** [comment]
Alright, I pinned the problem down to the app locking to an outdated version of kareem, which is a dependency of mongoose. I have no idea why it's doing that though... I'll have more time to look into it later in the week.

**Comment by modermo** [comment]
I just ran into this myself... basically followed the standard install.

**Comment by mkilp** [comment]
I ran into this problem aswell. 
Fixed it by ensuring that mongoose is connected to my database before creating the brute storage.

**Comment by filipporampado** [comment]
Having the same problem. Checking if mongoose is connected does not work for me. Changing BruteForceSchema to the one provided in the readme and not "express-brute-mongoose/dist/schema", works.

**Comment by n0bit** [comment]
I am getting this error. I tried waiting for mongoose to connect and changing the default schema to readme schema, but error persist.

Edit:
It works for me now. I had to make a new schema from scratch using the template in readme.

**Comment by noamichael** [comment]
I created pull request #14 which should hopefully fix this issue. As a temporary workaround, you can just create the schema with the correct mongoose version yourself:

```javascript
const ExpressBrute = require('express-brute');
const MongooseStore = require('express-brute-mongoose');
const BruteForceSchema = require('express-brute-mongoose/dist/schema');
const mongoose = require('mongoose');

const bruteForceSchema = new mongoose.Schema(//use our mongoose instance to create the schema
    {
        _id: { type: String },
        data: {
            count: Number,
            lastRequest: Date,
            firstRequest: Date
        },
        expires: { type: Date, index: { expires: '1d' } }
    },
    { collection: 'bruteforce' }
);

const model = mongoose.model('bruteforce', bruteForceSchema);

const store = new MongooseStore(model);

module.exports = new ExpressBrute(store);
```

**Comment by noamichael** [comment]
Another thought I had is that maybe the `schema.js` file should simply export the raw JSON of the schema instead, and we can completely remove mongoose as a dependency and let the user import whichever version they want to use:


```javascript
const ExpressBrute = require('express-brute');
const MongooseStore = require('express-brute-mongoose');
const BruteForceSchema = require('express-brute-mongoose/dist/schema');
const mongoose = require('mongoose');

const model = mongoose.model('bruteforce', new mongoose.Schema(BruteForceSchema));

const store = new MongooseStore(model);

module.exports = new ExpressBrute(store);

```

**Comment by MohammedAl-Mahdawi** [comment]
Same issue here, fixed it by replacing 

```
var BruteForceSchema = require('express-brute-mongoose/dist/schema');
```

with 

```
var BruteForceSchema = {
  "_id": String,
  "data": {
    "count": Number,
    "lastRequest": Date,
    "firstRequest": Date
  },
  "expires": Date
}
```

**Comment by cbargren** [comment]
Thanks for all your guys' input (and sorry for the radio silence). I've implemented @noamichael's suggestion of removing the mongoose dependency from the package altogether and just expect the user to create the schema themselves using a raw json object exported by the package via `express-brute-mongoose/dist/schema`. That's now live as of [v1.1.0](https://www.npmjs.com/package/express-brute-mongoose/v/1.1.0). Thanks and open a new issue if anyone's still having problems. 🍻
